src/components/tl/mlx5/mcast/tl_mlx5_mcast_helper.h

```diff
@@ -0,0 +1,322 @@
+            pp         = (struct pp_packet*) (id);
+            pp->length = wc[i].byte_len - GRH_LENGTH;
+            pp->psn    = ntohl(wc[i].imm_data);
+
+            tl_trace(comm->lib, "completion: psn %d, length %d, already_received %d, "
+                                " psn in req %d, req_start %d, req_num packets"
+                                " %d, to_send %d, to_recv %d, num_left %d",
+                                pp->psn, pp->length, PSN_RECEIVED(pp->psn,
+                                comm) > 0, PSN_IS_IN_RANGE(pp->psn, req,
+                                comm), req->start_psn, req->num_packets,
+                                req->to_send, req->to_recv, num_left);
+
+            if (PSN_RECEIVED(pp->psn, comm) || pp->psn < comm->last_acked) {
+                /* This psn was already received */
+                ucc_assert(pp->context == 0);
+                ucc_list_add_tail(&comm->bpool, &pp->super);
+            } else {
+                if (num_left > 0 && PSN_IS_IN_RANGE(pp->psn, req, comm)) {
+                    __builtin_prefetch(req->ptr + PSN_TO_RECV_OFFSET(pp->psn, req, comm));
+                    __builtin_prefetch((void*) pp->buf);
+                    ucc_tl_mlx5_mcast_process_packet(comm, req, pp);
+                    num_left--;
+                } else {
+                    ucc_list_add_tail(&comm->pending_q, &pp->super);
+                }
+            }
+        }
+
+        comm->pending_recv -= num_comp;
+        ucc_tl_mlx5_mcast_post_recv_buffers(comm);
+    }
+
+    ucc_free(wc);
+    return num_left;
+}
+
+/* this function returns the number of mcast recv packets
+ * that have been completed  or -1 in case of error */
+static inline int ucc_tl_mlx5_mcast_poll_recv(ucc_tl_mlx5_mcast_coll_comm_t *comm)
+{
+    struct pp_packet *pp;
+    struct ibv_wc     wc;
+    int               num_comp;
+    uint64_t          id;
+    int               length;
+    uint32_t          psn;
+
+    do {
+        num_comp = ibv_poll_cq(comm->rcq, 1, &wc);
+
+        if (num_comp > 0) {
+            
+            if (IBV_WC_SUCCESS != wc.status) {
+                tl_error(comm->lib, "mcast_poll_recv: %s err %d num_comp",
+                        ibv_wc_status_str(wc.status), num_comp);
+                return -1;
+            }
+
+            // Make sure we received all in order.
+            id     = wc.wr_id;
+            length = wc.byte_len - GRH_LENGTH;
+            psn    = ntohl(wc.imm_data);
+            pp     = (struct pp_packet*) id;
+
+            if (psn >= comm->psn) {
+                ucc_assert(!PSN_RECEIVED(psn, comm));
+                pp->psn    = psn;
+                pp->length = length;
+                ucc_list_add_tail(&comm->pending_q, &pp->super);
+            } else {
+                ucc_assert(pp->context == 0);
+                ucc_list_add_tail(&comm->bpool, &pp->super);
+            }
+
+            comm->pending_recv--;
+            ucc_tl_mlx5_mcast_post_recv_buffers(comm);
```

Comment:
check return value. Is it really needed to post next recv right after first completion? Maybe it's better to do it just before returning from this function to post all recv wqes at once?

Reply:
```ucc_tl_mlx5_mcast_post_recv_buffers``` first checks if the number of free recv WQs in the pool is less than ```post_recv_thres```. If it goes higher than the threshold, then we post at least ```post_recv_thres``` recv WQ to the QP. Therefore, this function most of the time just translates into a single if check and returns so that we are sure we are not draining the QP and we are always ready for receive (even if some interrupt happens during this while loop)